File: src/main/java/com/busted_moments/client/features/war/WeeklyWarCountOverlay.java
package com.busted_moments.client.features.war;

import com.busted_moments.client.models.war.events.WarStartEvent;
import com.busted_moments.client.util.ChatUtil;
import com.busted_moments.core.Default;
import com.busted_moments.core.Feature;
import com.busted_moments.core.State;
import com.busted_moments.core.config.Config;
import com.busted_moments.core.render.overlay.Hud;
import com.busted_moments.core.time.Duration;
import com.busted_moments.core.time.ChronoUnit;
import com.busted_moments.core.util.NumUtil;
import com.mojang.blaze3d.platform.Window;
import com.mojang.blaze3d.vertex.PoseStack;
import com.wynntils.core.components.Managers;
import com.wynntils.core.consumers.overlays.OverlayPosition;
import com.wynntils.mc.event.TickEvent;
import com.wynntils.utils.render.type.HorizontalAlignment;
import com.wynntils.utils.render.type.TextShadow;
import com.wynntils.utils.render.type.VerticalAlignment;
import me.shedaniel.math.Color;
import net.minecraft.ChatFormatting;
import net.minecraftforge.eventbus.api.SubscribeEvent;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Default(State.ENABLED)
@Config.Category("War")
@Feature.Definition(name = "Display Weekly War Count")
public class WeeklyWarCountOverlay extends Feature {
   private static final Duration WEEK = Duration.of(1, ChronoUnit.WEEKS);

   @Hud
   private static WeeklyWars HUD;

   @Value("Text Style")
   private static TextShadow style = TextShadow.OUTLINE;

   @Value("Text Color")
   private static Color text_color = ChatUtil.colorOf(ChatFormatting.LIGHT_PURPLE);

   @Alpha
   @Value("Background Color")
   private static Color background_color = Color.ofRGBA(0, 0, 0, 127);

   @Hidden("wars")
   private static List<Date> WARS = new ArrayList<>();

   private static int WEEKLY_WARS = 0;

   @SubscribeEvent
   private static void onWarStart(WarStartEvent event) {
      Managers.TickScheduler.scheduleNextTick(() -> WARS.add(new Date()));
   }

   @SubscribeEvent
   private static void onTick(TickEvent event) {
      WEEKLY_WARS = (int) WARS.stream()
              .filter(date -> Duration.since(date).lessThan(WEEK))
              .count();
   }

   @Hud.Name("Weekly Wars")
   @Hud.Offset(x = -250.0F, y = 0F)
   @Hud.Size(width = 91.5F, height = 14.415209F)
   @Hud.Anchor(OverlayPosition.AnchorSection.BOTTOM_MIDDLE)
   @Hud.Align(vertical = VerticalAlignment.BOTTOM, horizontal = HorizontalAlignment.RIGHT)
   private static class WeeklyWars extends Hud.Element {
      @Override
      protected void onRender(float x, float y, float width, float height, PoseStack poseStack, float partialTicks, Window window) {
         render(x, y, WEEKLY_WARS);
      }

      @Override
      protected void onRenderPreview(float x, float y, float width, float height, PoseStack poseStack, float partialTicks, Window window) {
         render(x, y, 13);
      }

      private void render(float x, float y, int wars) {
         if (wars == 0) return;

         new TextBox(builder -> builder
                 .append(NumUtil.format(wars))
                 .append(" War" + ((wars != 1) ? "s" : "")), x, y)
                 .setTextColor(text_color)
                 .setTextStyle(style)
                 .setFill(background_color)
                 .with(this)
                 .setPadding(5, 4, 5, 4)
                 .setMaxWidth(getWidth())
                 .dynamic()
                 .build();
      }
   }

   public static List<Date> getWars() {
      return WARS;
   }
}

File: src/main/kotlin/com/busted_moments/client/features/war/WeeklyWarCountOverlay.kt
package com.busted_moments.client.features.war

import com.busted_moments.client.framework.wynntils.esl
import com.busted_moments.client.framework.config.annotations.Category
import com.busted_moments.client.framework.config.entries.value.Value
import com.busted_moments.client.framework.events.Subscribe
import com.busted_moments.client.framework.features.AbstractOverlay
import com.busted_moments.client.framework.features.Align
import com.busted_moments.client.framework.features.At
import com.busted_moments.client.framework.features.Define
import com.busted_moments.client.framework.features.Feature
import com.busted_moments.client.framework.features.Overlays
import com.busted_moments.client.framework.features.Size
import com.busted_moments.client.framework.render.elements.textbox
import com.busted_moments.client.framework.text.Text
import com.busted_moments.client.framework.util.Numbers.escapeCommas
import com.busted_moments.client.framework.util.Numbers.toCommaString
import com.busted_moments.client.models.territories.war.WarModel
import com.wynntils.core.consumers.overlays.OverlayPosition
import com.wynntils.mc.event.TickEvent
import com.wynntils.utils.colors.CustomColor
import com.wynntils.utils.render.type.HorizontalAlignment
import com.wynntils.utils.render.type.TextShadow
import com.wynntils.utils.render.type.VerticalAlignment
import net.essentuan.esl.Rating
import net.essentuan.esl.color.Color
import net.essentuan.esl.model.annotations.Sorted
import net.essentuan.esl.time.duration.Duration
import net.essentuan.esl.time.duration.days
import net.essentuan.esl.time.extensions.timeSince
import net.minecraft.ChatFormatting


@Category("War")
@Overlays(WeeklyWarCountOverlay.Overlay::class)
object WeeklyWarCountOverlay : Feature() {
    @Value("Past")
    @Sorted(Rating.LOWEST)
    private var past: Duration = 7.days

    @Define(
        name = "Weekly War Overlay",
        at = At(
            x = -250f,
            y = 0f
        ),
        size = Size(
            width = 91.5f,
            height = 14.415209f
        ),
        anchor = OverlayPosition.AnchorSection.BOTTOM_MIDDLE,
        align = Align(
            vert = VerticalAlignment.BOTTOM,
            horizontal = HorizontalAlignment.RIGHT
        )
    )
    object Overlay : AbstractOverlay() {
        @Value("Text Style")
        private var style: TextShadow = TextShadow.OUTLINE

        @Value("Text Color")
        private var textColor: Color = CustomColor.fromInt(ChatFormatting.LIGHT_PURPLE.color!!).esl

        @Value("Background Color", alpha = true)
        private var backgroundColor: Color = CustomColor(0, 0, 0, 127).esl

        private var wars: Int = 0
        
        override fun render(ctx: Context): Boolean {
            textbox {
                frame()
                
                style = this@Overlay.style
                background = backgroundColor
                padding.all(5f)
                
                text = Text {
                    +wars.toCommaString().escapeCommas().color(textColor)
                    +" War"

                    if (wars != 1)
                        +"s"
                }
            }

            return true
        }

        @Subscribe
        private fun TickEvent.on() {
            wars = WarModel.count { it.started.timeSince() < past }
        }
    }
}